File: websocket/ws_client.js
const WebSocket = require('ws');
const CryptoJS = require('crypto-js'); // crypto-js@4.1.1


class FXClient {
  constructor(url, path, ssl, apiKey, passphrase, secretKey, onConnectCallback) {
    this.url = url;
    this.path = path;
    this.ssl = ssl;
    this.onConnectCallback = onConnectCallback;
    this.apiKey = apiKey;
    this.passphrase = passphrase;
    this.secretKey = secretKey;
    this.connection = null;
  }

  createAuthRequest() {
    const timestamp = parseInt(new Date().getTime() / 1000);
    const message = timestamp + 'GET' + this.path;
    const hmacKey = CryptoJS.enc.Base64.parse(this.secretKey);
    const signature = CryptoJS.HmacSHA256(message, hmacKey);
    const signatureB64 = signature.toString(CryptoJS.enc.Base64);
    return {
      'action': 'auth',
      'request_id': "my_auth_request_0",
      'sign': signatureB64,
      'timestamp': timestamp,
      'api_key': this.apiKey,
      'passphrase': this.passphrase,
    }
  }

  onConnect() {
    console.log(`${new Date()} Connected to websocket server`);
    this.authenticate()
  }

  authenticate() {
    const authRequest = this.createAuthRequest()
    this.connection.send(JSON.stringify(authRequest))
  }

  onError(msg) {
    console.log(`Error. Message received:`, msg)
  }

  onClose(msg) {
    console.log(`${new Date()}Lol. Close:`, msg)
  }

  connect() {
    if (this.connection) {
    }
    const protocol = this.ssl ? "wss" : "ws"
    const finalUrl = `${protocol}://${this.url}${this.path}`
    console.log(finalUrl)
    this.connection = new WebSocket(finalUrl);
    this.connection.on('open', this.onConnect.bind(this));
    this.connection.on('close', this.onClose.bind(this));
    this.connection.on('end', this.onClose.bind(this));
    this.connection.on('error', this.onError.bind(this));
    this.connection.on('message', this.onMessage.bind(this));
    this.connection.onclose = function (ev) {
      console.info(`${new Date()}Websocket closed.`);
    };
  }

  onMessage(msg) {
    const stringData = String.fromCharCode.apply(null, msg)
    const jsonMessage = JSON.parse(stringData)
    switch(jsonMessage.event) {
      case "auth_response": {
        if(jsonMessage.status == "success"){
          // Subscribe
          // this.subscribe()

          // Fetch data
          this.fetchData("max_levels")
          this.fetchData("allowed_markets")
          this.fetchData("max_connections")
        }else{
          console.log("Authentication Failure")
        }
        break
      }
      case "subscribe_response": {
        if(jsonMessage.status == "success"){
          console.log("Subscription successful  -> ", jsonMessage)
        }else{
          console.log("Subscription Failed -> ", jsonMessage)
        }
        break
      }
      case "unsubscribe_response": {
        if(jsonMessage.status == "success"){
          console.log("UnSubscription successful  -> ", jsonMessage)
        }else{
          console.log("UnSubscription Failed -> ", jsonMessage)
        }
        break
      }
      case "data_response": {
        if(jsonMessage.status == "success"){
          console.log("Data Response  -> ", JSON.stringify(jsonMessage,null,2))
        }else{
          console.log("Data Request Failed -> ", jsonMessage)
        }
        break
      }
      case "stream": {
        if(jsonMessage.status == "success"){
          console.log(JSON.stringify(jsonMessage,null,2))
        }else{
          console.log("Error in stream: ", JSON.stringify(jsonMessage,null,2))
        }
        break
      }
      case "error_response": {
        if(jsonMessage.status == "success"){
          console.log(JSON.stringify(jsonMessage,null,2))
        }else{
          console.log("Error in stream: ", JSON.stringify(jsonMessage,null,2))
        }
        break
      }
    }
  }

  subscribe(){
    const subscription_request = {
      "base_token": "ETH",
      "quote_token": "USD",
      "quantity": {
        "token": "ETH",
        "levels": [0.1, 1]
      },
      "request_id": "my_request_1",
      "action": "subscribe"
    }

    this.connection.send(JSON.stringify(subscription_request));
  }

  unsubscribe(){
    const subscription_request = {
      "base_token": "ETH",
      "quote_token": "USD",
      "request_id": "my_request_2",
      "action": "unsubscribe"
    }

    this.connection.send(JSON.stringify(subscription_request));
  }

  fetchData(requestTyoe){
    const subscription_request = {
      "request_type": requestTyoe,
      "request_id": "my_request_3",
      "action": "data_request"
    }

    this.connection.send(JSON.stringify(subscription_request));
  }

  emit(event, message) {
    if (!this.connection) {
      throw "connection is not initialised yet"
    }
    this.connection.emit(event, message);
  }
}

const url = "stream.falconx.io"
const path = "/price.tickers"

apiKey = "xxx"
secretKey = "xxx"
passphrase = "xxx"

var fxStreamingClient = new FXClient(url, path, true, apiKey, passphrase, secretKey);

fxStreamingClient.connect();

File: websocket/python/ws_client.py
import base64
import hashlib
import hmac
import json
import time
from pprint import pprint

import websocket
from marshmallow import fields
from marshmallow.validate import Range


class ConnectionOpts:
    retry_on_error = fields.Boolean(default=False)
    num_retries = fields.Integer(validate=Range(min=1, error="Value must be greater than 0"))
    retry_delay = fields.Integer()
    log_streams = fields.Boolean(default=False)

    def __init__(self,
                 retry_on_error=False,
                 num_retries=None,
                 retry_delay=None,
                 log_streams=False,
                 ) -> None:
        self.retry_on_error = retry_on_error
        self.num_retries = num_retries
        self.retry_delay = retry_delay
        self.log_streams = log_streams


class FalconXWSClient:
    host = fields.String(required=True)
    path = fields.String()
    ssl = fields.Boolean(default=False)
    api_key = fields.String(required=True)
    secret = fields.String(required=True)
    passphrase = fields.String(required=True)
    connection_opts = fields.Nested(ConnectionOpts)
    authenticated = fields.Boolean(default=False)
    reader_active = fields.Boolean(default=False)
    conn = fields.Nested(websocket.WebSocketApp)
    retry_count = 0

    def __init__(self,
                 host: str = None,
                 path: str = "/",
                 ssl: bool = False,
                 api_key: str = None,
                 secret: str = None,
                 passphrase: str = None,
                 connection_opts: ConnectionOpts = None
                 ) -> None:
        self.ssl = ssl
        self.path = path
        self.host = ("wss://" if ssl else "ws://") + host
        self.api_key = api_key
        self.secret = secret
        self.passphrase = passphrase
        self.connection_opts = connection_opts

    def get_signature(self, timestamp, path):
        method = "GET"
        message = "".join([timestamp, method, path, ""])
        hmac_key = base64.b64decode(self.secret)
        signature = hmac.new(hmac_key, message.encode(), hashlib.sha256)
        signature_b64 = base64.b64encode(signature.digest())
        signature = signature_b64.decode("utf-8")
        return signature

    def connect(self):
        def on_message(ws, message):
            # print(message)
            data = json.loads(message)
            if data["event"] == "auth_response":
                if data["status"] == "success":
                    print("Authentication successful.", data)  # noqa
                    self.authenticated = True

                    # Subscribe
                    self.subscribe()

                    # Fetch data
                    # self.fetch_data("max_levels")
                    # self.fetch_data("allowed_markets")
                    # self.fetch_data("max_connections")
                else:
                    print("Authentication Unsuccessful. Rsponse: ", data)  # noqa
            if data["event"] == "subscribe_response":
                if data["status"] == "success":
                    print("Subscription successful. Response: ", data)  # noqa
                else:
                    print("Subscription failed. Response: ", data)  # noqa
            if data["event"] == "unsubscribe_response":
                if data["status"] == "success":
                    print("UnSubscription successful. Response: ", data)  # noqa
                else:
                    print("UnSubscribe failed. Response: ", data)  # noqa
            if data["event"] == "data_response":
                if data["status"] == "success":
                    print("Data Response: ", data)  # noqa
                else:
                    print("Data request error: ", data)  # noqa
            if data["event"] == "error_response":
                print("Error Response: ", data)  # noqa
            elif data["event"] == "stream":
                if data["status"] == "success":
                    pprint(data)  # noqa
                else:
                    print("Error in stream: ", data)  # noqa

        def on_error(ws, error):
            print("Error : ", error)  # noqa

        def on_close(ws, close_status_code, close_msg):
            print("### closed ###")  # noqa

        def on_open(ws):
            print("Opened connection")  # noqa
            self.authenticate()

        # websocket.enableTrace(True)
        self.conn = websocket.WebSocketApp(self.host + self.path,
                                           on_open=on_open,
                                           on_message=on_message,
                                           on_error=on_error,
                                           on_close=on_close)
        self.conn.run_forever(reconnect=5)

    def authenticate(self):
        timestamp = int(time.time())
        signature = self.get_signature(str(timestamp), self.path)

        req = {
            "action": "auth",
            "api_key": self.api_key,
            "passphrase": self.passphrase,
            "signature": signature,
            "timestamp": timestamp,
            "request_id": "my_request"
        }
        self.conn.send(data=json.dumps(req))

    def subscribe(self):
        subscription_request = {
            "base_token": "ETH",
            "quote_token": "USD",
            "quantity": {
                "token": "ETH",
                "levels": [0.1, 1]
            },
            "request_id": "my_request_1",
            "action": "subscribe"
        }
        self.conn.send(data=json.dumps(subscription_request))

    def unsubscribe(self, base_token, quote_token):
        unsubscription_request = {
            "base_token": base_token,
            "quote_token": quote_token,
            "request_id": "my_request_1",
            "action": "unsubscribe"
        }
        self.conn.send(data=json.dumps(unsubscription_request))

    def fetch_data(self, type_of_data):
        data_request = {
            "request_type": type_of_data,
            "request_id": "my_request_1",
            "action": "data_request"
        }
        self.conn.send(data=json.dumps(data_request))


if __name__ == "__main__":
    path = "/price.tickers"
    host = "stream.falconx.io"
    api_key = 'xxx'
    passphrase = "xxx"
    secret_key = "xxx"

    connection_opts = ConnectionOpts(retry_on_error=True, num_retries=5, retry_delay=1)

    fx_ws_client = FalconXWSClient(host=host, path=path, api_key=api_key, secret=secret_key, passphrase=passphrase, connection_opts=connection_opts, ssl=True)

    fx_ws_client.connect()
